### PR TITLE
TagManager and Tags

### PR DESCRIPTION
**Pull request by fterh**
In progress, don't merge yet.

### CODE REVIEW

**Comment by fterh** [comment]
Ready to merge.

Important: review the API (TagManagerInterface.swift) and see if it's okay :) Implementation and tests follow the interface specifications.

**Comment by fterh** [comment]
Ready to merge.

Main features:
* TagManager is implemented as semi-singleton with private init()
* To instantiate, use static method `.create(withPersistence: Bool)`
* Non-persistent TagManager for unit testing (of logic)
* Persistent TagManager (singleton) for production
* TagManager automatically saves its state to storage (persistence)
* When creating a persistent TagManager, try reconstructing from saved data (if possible); otherwise, create an empty/fresh instance of a persistent TagManager
* Logging for I/O methods (since we can't test it)

File: bacon/model/Tags/TagManager.swift [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+    func addChildTag(_ child: String, to parent: String) throws {
+        let childTag = Tag(child, parent: parent)
+        let parentTag = Tag(parent)
+
+        // childTag should not already exist
+        guard !allTags.contains(childTag) else {
+            throw DuplicateTagError(message: "\(childTag) already exists")
+        }
+
```
Comment (kumuwu) [inline bacon/model/Tags/TagManager.swift:85]: Shall assert the parent tag already exists?

File: bacon/model/Tags/TagManager.swift [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        // parentTag should exist
+        guard allTags.contains(parentTag) else {
+            throw InvalidTagError(message: "\(parentTag) does not exist")
+        }
+
+        removeTags([parentTag])
+    }
+
+    var tags: [Tag: Set<Tag>] {
+        return parentChildMap
```
Comment (kumuwu) [inline bacon/model/Tags/TagManager.swift]: I think I might need you to return me a a dictionary where the value is an array of child tags; from the UI side there might need to be a consistent ordering of the child tags under a parent tag

File: bacon/model/Tags/TagManager.swift [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    func getChildrenTagsOf(_ parent: String) throws -> Set<Tag> {
+        let parentTag = Tag(parent)
```
Comment (kumuwu) [inline bacon/model/Tags/TagManager.swift:155]: same here, an (sorted) array might be better

File: bacon/model/Tags/TagManager.swift (hunk 1 above)
Comment (kumuwu) [reply bacon/model/Tags/TagManager.swift:85]: I see that in the later part there is a method to add the parent as well when the parent is not already there. From the user side, he should never be able to add a child tag to a non-existent parent tag tho? if it happens it's probably some error within the programme?

Comment (fterh) [reply bacon/model/Tags/TagManager.swift:85]: Okay, I'll change the specifications then. 

Edit: fixed in commit 81bba2

File: bacon/model/Tags/TagManager.swift (hunk 2 above)
Comment (fterh) [reply bacon/model/Tags/TagManager.swift]: This is by design. Within TagManager, I need to do membership check/add/remove operations efficiently, but TagManager does not maintain tags in any order. So an array doesn't make sense.

I think it'll be better if you convert to an array then sort haha

Comment (kumuwu) [reply bacon/model/Tags/TagManager.swift]: As in, you can and should still internally maintain as a set, which will be unknown to me. I'm not sure if I should do any operations (i.e. sorting) on the tags? (separation of concern blah blah)
it's just like transactions internally are not stored in chronological order, but when UI gets hold of it it gets a sorted copy.
but, I'm not sure haha

Comment (fterh) [reply bacon/model/Tags/TagManager.swift]: Okay, done in commit 67f01e5 :)

https://github.com/cs3217-crispybacon/bacon/pull/60/commits/67f01e532c40463a15b253a6ea0d00eb30e2bbfc

**Review by kumuwu — APPROVED** [review]

**Pull request merged**